pkg/gktest/read_suites.go

```diff
@@ -1,18 +1,175 @@
+// 2) If the path is a directory and recursive is false, returns only the Suites
+//    defined in that directory.
+// 3) If the path is a directory and recursive is true returns all Suites in that
 //      directory and its subdirectories.
 //
 // Returns an error if:
 // - path is a file that does not define a Suite
 // - any matched files containing Suites are not parseable
-func ReadSuites(f fs.FS, path string) ([]Suite, error) {
-	return nil, nil
+func ReadSuites(f fs.FS, target string, recursive bool) ([]Suite, error) {
+	if f == nil {
+		return nil, ErrNoFileSystem
+	}
+	if target == "" {
+		return nil, ErrNoTarget
+	}
+
+	stat, err := fs.Stat(f, target)
+	if err != nil {
+		return nil, err
+	}
+
+	files := fileList{}
+	switch {
+	case !stat.IsDir() && !recursive:
+		// target is a file.
+		err = files.addFile(target)
+
+	case !stat.IsDir() && recursive:
+		// target is a file, but user specified it should be traversed recursively.
+		err = fmt.Errorf("can only recursively traverse directories, %w: %q",
+			ErrNotADirectory, target)
+
+	case stat.IsDir() && !recursive:
+		// target is a directory, but user did not specify to test subdirectories.
+		err = files.addDirectory(f, target)
+
+	case stat.IsDir() && recursive:
+		// target is a directory, and user specified it should be traversed recursively.
+		err = fs.WalkDir(f, target, files.walkEntry)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return readSuites(f, files)
+}
+
+// readSuites reads the passed set of files into Suites on the given filesystem.
+func readSuites(f fs.FS, files []string) ([]Suite, error) {
+	var suites []Suite
+	for _, file := range files {
+		suite, err := readSuite(f, file)
+		if err != nil {
+			return nil, err
+		}
+		if suite != nil {
+			suites = append(suites, *suite)
+		}
+	}
+	return suites, nil
+}
+
+// fileList is a convenience type for breaking apart and deduplicating code
+// related to collecting the set of files which may contain Suites.
+type fileList []string
+
+func (l *fileList) addFile(target string) error {
+	// target is a file.
+	ext := filepath.Ext(target)
+	if ext != ".yaml" && ext != ".yml" {
+		return fmt.Errorf("%w: %q", ErrUnsupportedExtension, ext)
+	}
+	*l = append(*l, target)
+	return nil
+}
+
+func (l *fileList) addDirectory(f fs.FS, target string) error {
+	// target is a directory, but user did not specify to test subdirectories.
+	entries, err := fs.ReadDir(f, target)
+	if err != nil {
+		return err
+	}
+	for _, entry := range entries {
+		err = l.walkEntry(filepath.Join(target, entry.Name()), entry, nil)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (l *fileList) walkEntry(path string, d fs.DirEntry, err error) error {
+	if err != nil {
+		return err
+	}
+	if isTestFile(d) {
+		*l = append(*l, path)
+	}
+	return nil
+}
+
+func isTestFile(d fs.DirEntry) bool {
+	if d.IsDir() {
+		return false
+	}
+	ext := filepath.Ext(d.Name())
+	return ext == ".yaml" || ext == ".yml"
```

Comment:
Maybe note that the test for whether a YAML is actually a test file by reading the contents is done elsewhere?

Reply:
Renamed to "isYAMLFile()"